src/field_tasks.c

```diff
@@ -0,0 +1,292 @@
+#include "global.h"
+#include "bike.h"
+#include "event_data.h"
+#include "field_camera.h"
+#include "field_effect_helpers.h"
+#include "field_player_avatar.h"
+#include "fieldmap.h"
+#include "metatile_behavior.h"
+#include "overworld.h"
+#include "quest_log.h"
+#include "script.h"
+#include "sound.h"
+#include "task.h"
+#include "constants/flags.h"
+#include "constants/songs.h"
+
+static void DummyPerStepCallback(u8 taskId);
+static void AshGrassPerStepCallback(u8 taskId);
+static void IcefallCaveIcePerStepCallback(u8 taskId);
+static void CrackedFloorPerStepCallback(u8 taskId);
+
+static const TaskFunc sPerStepCallbacks[] =
+{
+    DummyPerStepCallback,
+    AshGrassPerStepCallback,
+    DummyPerStepCallback,
+    DummyPerStepCallback,
+    IcefallCaveIcePerStepCallback,
+    DummyPerStepCallback,
+    DummyPerStepCallback,
+    CrackedFloorPerStepCallback
+};
+
+static const u8 sIcefallCaveIceTileCoords[][2] =
+{
+    { 0x08, 0x03 },
+    { 0x0a, 0x05 },
+    { 0x0f, 0x05 },
+    { 0x08, 0x09 },
+    { 0x09, 0x09 },
+    { 0x10, 0x09 },
+    { 0x08, 0x0a },
+    { 0x09, 0x0a },
+    { 0x08, 0x0e }
+};
+
+static void Task_RunPerStepCallback(u8 taskId)
+{
+    int idx = gTasks[taskId].data[0];
+    sPerStepCallbacks[idx](taskId);
+}
+
+static void Task_RunTimeBasedEvents(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+
+    if (!ScriptContext2_IsEnabled())
+    {
+        if (gUnknown_203ADFA != 2 && gUnknown_203ADFA != 3)
+        {
+            sub_8056078(&data[1], &data[2]);
+        }
+    }
+}
+
+void SetUpFieldTasks(void)
+{
+    if (!FuncIsActiveTask(Task_RunPerStepCallback))
+    {
+        u8 taskId = CreateTask(Task_RunPerStepCallback, 0x50);
+        gTasks[taskId].data[0] = 0;
+    }
+
+    if (!FuncIsActiveTask(Task_RunTimeBasedEvents))
+        CreateTask(Task_RunTimeBasedEvents, 0x50);
+}
+
+void ActivatePerStepCallback(u8 callbackId)
+{
+    u8 taskId = FindTaskIdByFunc(Task_RunPerStepCallback);
+    if (taskId != 0xff)
+    {
+        s32 i;
+        s16 *data = gTasks[taskId].data;
+
+        for (i = 0; i < 16; i++)
+            data[i] = 0;
+
+        if (callbackId >= NELEMS(sPerStepCallbacks))
+        {
+            data[0] = 0;
+        }
+        else
+        {
+            data[0] = callbackId;
+        }
+    }
+}
+
+void ResetFieldTasksArgs(void)
+{
+    u8 taskId;
+    s16 *data;
+
+    taskId = FindTaskIdByFunc(Task_RunPerStepCallback);
+    if (taskId != 0xff)
+    {
+        data = gTasks[taskId].data;
+    }
+    taskId = FindTaskIdByFunc(Task_RunTimeBasedEvents);
+    if (taskId != 0xff)
+    {
+        data = gTasks[taskId].data;
+        data[1] = 0;
+        data[2] = 0;
+    }
+}
+
+static void DummyPerStepCallback(u8 taskId)
+{
+}
+
+static void MarkIcefallCaveCoordVisited(s16 x, s16 y)
+{
+    u8 i = 0;
+    for (; i < NELEMS(sIcefallCaveIceTileCoords); ++i)
+    {
+        if (sIcefallCaveIceTileCoords[i][0] + 7 == x && sIcefallCaveIceTileCoords[i][1] + 7 == y)
```

Comment:
Is there a constant for the `+ 7` here and in `Special_SetIcefallCaveCrackedIceMetatiles`?

Reply:
Not that I'm aware of.